core/kernel/tee_ta_manager.c

```diff
@@ -823,6 +823,174 @@ TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
+		res = TEE_ERROR_BUSY;
+	}
+
+	return res;
+}
+
+TEE_Result tee_ta_instance_stats(void *buff,
+				 uint32_t *buff_size)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_param param = { };
+	struct tee_ta_dump_stats *dump_stats = NULL;
+	struct tee_ta_dump_ctx *dump_ctx = NULL;
+	struct tee_ta_session *sess = NULL;
+	struct tee_ta_session_head *open_sessions = NULL;
+	struct tee_ta_ctx *ctx = NULL;
+	uint32_t actual_size = 0;
+	uint32_t ta_instance_num = 0;
+	uint32_t dump_num = 0;
+
+	if (!buff_size)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mutex_lock(&tee_ta_mutex);
+
+	/*
+	 * Go through all available TA and calc out the actual buffer size.
+	 */
+	TAILQ_FOREACH(ctx, &tee_ctxes, link)
+		ta_instance_num++;
+
+	actual_size = sizeof(struct tee_ta_dump_stats) * ta_instance_num;
+
+	if (!buff || *buff_size == 0) {
+		/*
+		 * Buff is null or pass size with 0 means caller want to get
+		 * the size of buff.
+		 */
+		*buff_size = actual_size;
+		mutex_unlock(&tee_ta_mutex);
+		return res;
+	}
+
+	/*
+	 * If the buffer size is smaller than the actual size,
+	 * update *buff_size and return TEE_ERROR_SHORT_BUFFER
+	 */
+	if (*buff_size < actual_size) {
+		*buff_size = actual_size;
+		res = TEE_ERROR_SHORT_BUFFER;
+		mutex_unlock(&tee_ta_mutex);
+		return res;
+	}
+
+	dump_stats = (struct tee_ta_dump_stats *)buff;
+	actual_size = ta_instance_num * sizeof(struct tee_ta_dump_ctx);
+	dump_ctx = (struct tee_ta_dump_ctx *)malloc(actual_size);
+	nsec_sessions_list_head(&open_sessions);
+	/*
+	 * Scan all sessions opened from secure side by searching through
+	 * all available TA instances and for each context, scan all opened
+	 * sessions.
+	 */
+	/* Context of a loaded TA */
+	TAILQ_FOREACH(ctx, &tee_ctxes, link) {
+		uint32_t cnt = 0;
+
+		memcpy(&dump_stats[dump_num].uuid, &ctx->ts_ctx.uuid,
+		       sizeof(ctx->ts_ctx.uuid));
+		dump_stats[dump_num].panicked = ctx->panicked;
+		dump_stats[dump_num].is_user_ta =
+			is_user_ta_ctx(&ctx->ts_ctx);
+
+		memcpy(&dump_ctx[dump_num].uuid, &ctx->ts_ctx.uuid,
+		       sizeof(ctx->ts_ctx.uuid));
+		dump_ctx[dump_num].panicked = ctx->panicked;
+		dump_ctx[dump_num].is_user_ta =
+			is_user_ta_ctx(&ctx->ts_ctx);
+		TAILQ_FOREACH(sess, open_sessions, link) {
+			if (sess->ts_sess.ctx == &ctx->ts_ctx) {
+				if (cnt == MAX_DUMP_SESS_NUM)
+					break;
+
+				dump_ctx[dump_num].sess_id[cnt] =
+					sess->id;
+				cnt++;
+			}
+		}
+
+		dump_stats[dump_num].sess_num = cnt;
+		dump_ctx[dump_num].sess_num = cnt;
+		dump_num++;
+	}
+	mutex_unlock(&tee_ta_mutex);
+
+	/*
+	 * Go through the dump context to get heap stats if it's user.
+	 */
+	for (uint32_t i = 0; i < dump_num; i++) {
+		struct tee_ta_dump_stats *stats = &dump_stats[i];
+
+		if (!dump_ctx[i].is_user_ta || dump_ctx[i].panicked)
+			continue;
+
+		/*find a session from TA dump context*/
+		sess = NULL;
+		for (uint32_t j = 0; j < dump_ctx[i].sess_num; j++) {
+			uint32_t sess_id = dump_ctx[i].sess_id[j];
+
+			sess = tee_ta_get_session(sess_id,
```

Comment:
should we get the session while the mutex is locked, to ensure the session still belongs to what we found scanning locked `tee_ctxs` list line 933?

Reply:
My previous work is getting session and entering user ta during `tee_ta_mutex` locking. That brings one concern that we enter user ta with holding a lock and turn into a big critical section here. Following @jenswi-linaro suggestion, I move it out of the lock and uses `tee_ta_get_session(sess_id)`/`tee_ta_put_session(sess)` here. I found that those two function are maintaining session's `ref_count` and that can ensure we are safe to access session context without holding the lock.